spec/lib/reporting/protocols_report_spec.rb

```diff
@@ -0,0 +1,172 @@
+require 'rails_helper'
+require 'reporting/protocols_report'
+
+RSpec.describe Reporting::ProtocolsReport do
+  let(:time_range) { Date.new(2022, 1, 1).all_day }
+
+  subject(:report) { Reporting::ProtocolsReport.new(issuers: nil, time_range:) }
+
+  before do
+    cloudwatch_client = double(
+      'Reporting::CloudwatchClient',
+      fetch: [
+        {
+          'protocol' => 'SAML Auth',
+          'issuer' => 'Issuer1',
+          'request_count' => '8',
+          'unsigned_count' => '3',
+          'invalid_signature_count' => '3',
+        },
+        {
+          'protocol' => 'SAML Auth',
+          'issuer' => 'Issuer2',
+          'request_count' => '2',
+          'unsigned_count' => '0',
+          'invalid_signature_count' => '0',
+        },
+        {
+          'protocol' => 'SAML Auth',
+          'issuer' => 'Issuer3',
+          'request_count' => '10',
+          'unsigned_count' => '4',
+          'invalid_signature_count' => '0',
+        },
+        {
+          'protocol' => 'OpenID Connect: authorization request',
+          'issuer' => 'Issuer4',
+          'request_count' => '80',
+          'unsigned_count' => '0',
+          'invalid_signature_count' => '0',
+        },
+      ],
+    )
+
+    allow(report).to receive(:cloudwatch_client).and_return(cloudwatch_client)
+  end
+
+  describe '#as_tables' do
+    it 'generates the tabular csv data' do
+      expect(report.as_tables).to eq expected_tables
+    end
+  end
+
+  describe '#as_emailable_reports' do
+    it 'adds a "first row" hash with a title for tables_report mailer' do
+      reports = report.as_emailable_reports
+      aggregate_failures do
+        reports.each do |report|
+          expect(report.title).to be_present
+        end
+      end
+    end
+  end
+
+  describe '#to_csvs' do
+    it 'generates a csv' do
+      csv_string_list = report.to_csvs
+      expect(csv_string_list.count).to be 3
+
+      csvs = csv_string_list.map { |csv| CSV.parse(csv) }
+
+      aggregate_failures do
+        csvs.map(&:to_a).zip(expected_tables(strings: true)).each do |actual, expected|
+          expect(actual).to eq(expected)
+        end
+      end
+    end
+  end
+
+  describe '#cloudwatch_client' do
+    let(:opts) { {} }
+    let(:subject) { described_class.new(issuers: [issuer], time_range:, **opts) }
+    let(:default_args) do
+      {
+        num_threads: 10,
+        ensure_complete_logs: true,
+        slice_interval: 1.day,
+        progress: false,
+        logger: nil,
+      }
+    end
+
+    describe 'when all args are default' do
+      xit 'creates a client with the default options' do
```

Comment:
I had disabled several of these tests as they were more tests of the cloudwatch reporting framework rather than _this_ report. Need an opinion:
1. Remove them entirely
2. Enable them any way
3. Refactor the framework so that the tests are not duplicated in each report.

Opinions?

Reply:
Tagging @Sgtpluck , @zachmargolis .

Reply:
i see what you're saying, but i see these more as testing what arguments we are passing through to the CloudwatchClient, rather than testing the framework itself. it's making sure that passing different options through the initializer leads to the correct arguments being passed to the framework.  

so i don't think they should be removed entirely -- abstracting out a `BaseReport` class if those methods are similar for all the existing reports is definitely an option, if it doesn't enlarge the scope of this too much